stl/src/atomic_ref.cpp

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// implement atomic_ref spin lock
+
+#include <atomic>
+#include <cstdint>
+#include <intrin.h>
+#include <new>
+#include <synchapi.h>
+#include <xthreads.h>
+
+_EXTERN_C
+_Smtx_t* __stdcall __std_atomic_get_mutex(const void* const _Key) noexcept {
+    constexpr size_t _Table_size_power = 8;
+    constexpr size_t _Table_size       = 1 << _Table_size_power;
+    constexpr size_t _Table_index_mask = _Table_size - 1;
+
+    struct alignas(std::hardware_destructive_interference_size) _Table_entry {
+        _Smtx_t _Mutex                                                           = 0;
+        char _Pad[std::hardware_destructive_interference_size - sizeof(_Smtx_t)] = {};
+    };
+    static _Table_entry _Table[_Table_size];

```

Comment:
Having a whole table for this might be overkill as it will only engage on systems that had at most 2 cores. My initial implementation back in 2015 only used one critical section shared by all 16 byte atomics; that might be OK here. This isn't wrong though.

Reply:
The table is for non-lock-free `atomic_ref`, not for a fallback.
Though it can be used for fallback too.

Reply:
🤦‍♂️ Makes sense, thanks!